README.md
# CSFinalProject
In the project, Scientific Python Interactive Data Acoustic Modeling, the program recieves an audio file, converts it into a .wav file and deletes the meta data that is not needed. After doing so, it offers the user options to plot the graph of apmlitude vs time, the highest frequencies, the lowest frequencies, the middle frequencies, and a graph of those three frequencies combined.

This project is meant to give a look at how different portions of the problem solving process can come together to solve a problem. This project gave us experiance with working as a team, designing a scrum board, working on coding tasks together and communicating with the other members in order to plan the process we decided to follow. The program itself helps to view data that could be useful in designing rooms in order to make the voice intelligibility better.

To use this program, open in pycharm. Then click the run button. After click the Browse button on the pop up and look for the file on your computer. After selecting the file click the Load button. Check the message at the bottom of the pop up to see if the load was succesful. After a successful load, click on the Plot Waveform button to plot the amplitude vs time. Click the Time button to see the time of the file. If you want to see the graph for the highest frequency click on the button labeled RT60 three times. Click the button labeled RT60 once for the low frequency. Click on the button labeled RT60 twice to graph middle frequencies. Click on the combined graph button to see the combined frequencies graph. Click the x in the top right to close the pop up. If you want to see the frequency for the whole file click the Plot Frequency button.


CHANGELOG.md
# CHANGELOG
# ChangeLog
## November 10, 2023

    On november 10, The repository in GitHub was created and a function was 
    added that was meant to convert any audiio file to a .wav file.

## November 13, 2023

    On November 13, the group members attempted to test how GitHub worked with pycharm and test if the converttowav() function worked. Functions for checked the file type and for saving the file
    path were added. Then the frame for the pop up box was created and files were merged. Then the created functions were placed into a single class. The graph was fixed so that it showed up in the
    same pop up instead of a new one.
    
## November 14, 2023

    On November 14, a button was added to collect the data from the given file along with a function that performs the action.
    
## November 15, 2023

    On November 15, a merge error occured and the file was rolled back. Then the code the creat a pop up window was placed in a class.

## November 17, 2023

   On November 17, new files were added to test the code.

## November 29, 2023

    On November 29, the function to display the length of timw a sound file is was completed.

## December 1, 2023 

    On December 1, the pop up window had a fram added to display the frequency values and the highest frequency was found.
    
## December 3, 2023

    On December 3, an attempt was made to retreive the lowest frequency and it had success.
        
## December 4, 2023

    On December 4 a README file was created. The graph for the highest frequency was made, and the graph for the lowest frequency was started.
    
## December 9, 2023
    On December 9, a requirement.txt was created. There was an attempt at placing the code into MVC format. 


test.py
import numpy as np
from pydub import AudioSegment
import matplotlib.pyplot as plt
from scipy.io import wavfile

def plot_specific_frequencies(audio_path, frequency_range):
    # Load the audio file
    audio = AudioSegment.from_file(audio_path)

    # Convert AudioSegment to NumPy array
    samples = np.array(audio.get_array_of_samples())

    # Perform Fourier Transform
    frequencies, amplitudes = calculate_frequency_spectrum(samples, audio.frame_rate)

    # Filter frequencies within the specified range
    mask = np.logical_and(frequencies >= frequency_range[0], frequencies <= frequency_range[1])
    filtered_frequencies = frequencies[mask]
    filtered_amplitudes = amplitudes[mask]

    # Plot the specific frequencies
    plt.figure(figsize=(10, 6))
    plt.plot(filtered_frequencies, filtered_amplitudes)
    plt.title(f"Frequency Spectrum ({frequency_range[0]} Hz to {frequency_range[1]} Hz)")
    plt.xlabel("Frequency (Hz)")
    plt.ylabel("Amplitude")
    plt.show()


def calculate_frequency_spectrum(samples, frame_rate):
    # Perform Fourier Transform
    fft_result = np.fft.fft(samples)

    # Calculate the frequencies corresponding to the FFT result
    frequencies = np.fft.fftfreq(len(fft_result), 1 / frame_rate)

    # Take the absolute value to get the amplitude spectrum
    amplitudes = np.abs(fft_result)

    # Keep only the positive frequencies
    positive_frequencies = frequencies[:len(frequencies) // 2]
    positive_amplitudes = amplitudes[:len(amplitudes) // 2]

    return positive_frequencies, positive_amplitudes


# Example usage
audio_path = "PolyHallClap_10mM.WAV"
frequency_range = (20, 50000)  # Specify the frequency range of interest
plot_specific_frequencies(audio_path, frequency_range)


view.py
import tkinter as tk
from tkinter import ttk
from tkinter import *
from tkinter import filedialog
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import matplotlib.pyplot as plt
from controller import Controller
from model import Model

class View(ttk.Frame):
    def __init__(self, parent):
        super().__init__(parent)

        self.controller = None
        self.status_msg = StringVar()

        # create widgets
        self.master.geometry('1000x700')
        self.master.rowconfigure(0, weight=1)
        self.master.columnconfigure(0, weight=1)

        self.mainframe = ttk.Frame(self.master, padding='5 5 5 5')
        self.mainframe.grid(row=0, column=0, sticky="NEWS")
        self.mainframe.rowconfigure(0, weight=1)
        self.mainframe.columnconfigure(0, weight=1)

        # filepath entry
        self.filepath_var = tk.StringVar()
        self.load_file_frame = ttk.LabelFrame(self.mainframe, padding='5 5 5 5', text='Choose File')
        self.load_file_frame.grid(column=0, row=0, sticky='NEW')
        self.load_file_frame.columnconfigure(0, weight=1)
        self.load_file_frame.rowconfigure(0, weight=1)

        choose_file_label = ttk.Label(self.load_file_frame, text='Please input file you would like to analyze:')
        choose_file_label.grid(column=0, row=0, columnspan=1, sticky='NEWS')

        self.choose_file_text = ttk.Entry(self.load_file_frame, width=70, textvariable=self.filepath_var)
        self.choose_file_text.grid(column=0, row=1, sticky='NEWS')

        choose_file_button = ttk.Button(self.load_file_frame, text="Browse", command=self.getfilepath)
        choose_file_button.grid(row=2, column=0, sticky='W')

        load_file_button = ttk.Button(self.load_file_frame, text="Load", command=self.load_button_clicked)
        load_file_button.grid(row=1, column=1, sticky='E')

        self.data_file_frame = ttk.LabelFrame(self.mainframe, padding='5 5 5 5', text='Data')
        self.data_file_frame.grid(column=0, row=2, sticky='NEW')
        self.data_file_frame.rowconfigure(1, weight=1)
        self.data_file_frame.columnconfigure(0, weight=1)

        show_plot = ttk.Button(self.data_file_frame, text='Plot', command=self.create_plot_call_to_controller)
        show_plot.grid(column=0, row=1, sticky='WN')

        # show_time = ttk.Button(self.data_file_frame, text='Time', command=self.extracttime)
        # show_time.grid(column=0, row=2, sticky='WN')
        #
        self.status_frame = ttk.Frame(self.master, relief='sunken', padding='2 2 2 2')
        self.status_frame.grid(row=1, column=0, sticky='EWS')
        self.status_msg.set('')
        status = ttk.Label(self.status_frame, textvariable=self.status_msg, anchor=W)
        status.grid(row=0, column=0, sticky='EW')

        # self.extra_data_file_frame = ttk.LabelFrame(self.mainframe, padding='5 5 5 5', text='Audio Data')
        # self.extra_data_file_frame.grid(column=0, row=1, sticky='NEW')
        # self.extra_data_file_frame.rowconfigure(1, weight=1)
        # self.extra_data_file_frame.columnconfigure(0, weight=1)
        #
        # self.highest_resonance.set(f'Frequency of Highest Amplitude: ')
        #
        # self.extra_data_highest_resonance_data = ttk.Label(self.extra_data_file_frame,
        #                                                    textvariable=self.highest_resonance)
        # self.extra_data_highest_resonance_data.grid(row=0, column=0, sticky='NW')
        #

    def set_controller(self, controller):
        self.controller = controller
    def getfilepath(self):
        self.filepath_var.set(tk.filedialog.askopenfilename())

    def load_button_clicked(self):
        if self.controller:
            self.controller.save(self.filepath_var.get())

    def create_plot_call_to_controller(self):
        self.controller.model.create_plot()

    def sb(self, msg):
        self.status_msg.set(msg)

model.py
from tkinter import StringVar
from tkinter import ttk
import numpy as np
from pydub import AudioSegment
import wave
import os
import tkinter as tk
from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
import matplotlib.pyplot as plt
from tkinter import filedialog
from scipy.io import wavfile
import re

class Model:

    def init(self, filepath):
        self.__filepath = filepath



        # to be extracted once load is called
        self.sample_rate = None
        self.length = None
        self.data = None
        self.spectrum = None
        self.freqs = None
        self.timedata = None
        self.im = None


    @property
    def filepath(self):
        return self.__filepath

    @filepath.setter
    def filepath(self, value):
        self.__filepath = value


    def create_plot(self):
        plt.plot(self.timedata, self.data[:, 0], label="Left Channel")
        plt.plot(self.timedata, self.data[:, 1], label="Right Channel")

        plt.legend()
        plt.xlabel('Time (s)')
        plt.ylabel('Amplitude')
        plt.show()

controller.py
import tkinter as tk
from tkinter import filedialog
import os

import matplotlib.pyplot as plt
import mutagen
import numpy as np
from pydub import AudioSegment
from scipy.io import wavfile


class Controller:

    def __init__(self, model, view):
        self.model = model
        self.view = view


    def save(self, filepath):
        try:
            # save the model
            self.model.filepath = filepath

            # show a success message
            self.view.sb(f'The filepath {filepath} saved!')
            self.convert_to_wav()

        except ValueError as error:
            # show an error message
            self.view.sb(error)

    def get_file_extension(self):
        return os.path.splitext(self.model.filepath)[-1]

    def get_file_name_with_extension(self):
        return self.model.filepath.get().split('/')[-1]

    def remove_metadata(self):
        metadata = mutagen.Metadata.load(self.model.filepath)
        metadata.delete()

    def convert_to_wav(self):
        if self.get_file_extension() == '.wav':
            # self.remove_metadata()
            self.model.sample_rate, self.model.data = wavfile.read(self.model.filepath)
            self.model.length = self.model.data.shape[0] / self.model.sample_rate
            self.model.timedata = np.linspace(0., self.model.length, self.model.data.shape[0])

        else:
            # self.remove_metadata()
            user_audio_file = AudioSegment.from_file(
                self.model.filepath,
                format=self.get_file_extension().strip('.')
            )
            # new_audio_file = AudioSegment(
            #     data=user_audio_file.raw_data,
            #     frame_rate=user_audio_file.frame_rate,
            #     sample_width=user_audio_file.sample_width,
            #     channels=user_audio_file.channels,
            #
            # )
            temp_wav_filepath = "temp.wav"

            os.system(f'ffmpeg -i {self.model.filepath} ./{temp_wav_filepath}')

            self.model.sample_rate, self.model.data = wavfile.read(temp_wav_filepath)
            self.model.spectrum, self.model.freqs, self.model.timedata, self.model.image = plt.specgram(
                self.model.data, Fs=self.model.sample_rate, NFFT=1024, cmap=plt.get_cmap('autumn_r')
            )

            os.remove(temp_wav_filepath)


    def check_for_wav(self):
        file_extension = self.get_file_extension()
        return file_extension == '.wav'



soundapp.py
import datetime
import os.path
import scipy
from tkinter import *
from tkinter import ttk, filedialog
import tkinter as tk
from matplotlib.figure import Figure
from pydub import AudioSegment
import wave
import matplotlib.pyplot as plt
from pydub.utils import mediainfo
from scipy.fft import fft
from scipy.signal import welch
from matplotlib.backends.backend_tkagg import (FigureCanvasTkAgg)
import numpy as np


class AudioAnalyzerApp:
    def __init__(self, master):
        self.master = master
        self.master.title("Audio Analyzer")
        self.mainframe = None
        self.load_file_frame = None
        self.status_frame = None
        self.choose_file_text = None
        self.wav_audio = None
        self.data_file_frame = None
        self.raw_data = None
        self.RT60values = None
        self.count = 0
        self.RT60values = {1: None, 2: None, 3: None}
        self.frame_amount = None
        self.highest_resonance = StringVar()

        self._filepath = StringVar()
        self.str_filepath = None
        self._status_msg = StringVar()

        self.create_widgets()

    def create_widgets(self):
        # master is essentially the same as root in tkinter
        self.master.geometry('1000x700')
        self.master.rowconfigure(0, weight=1)
        self.master.columnconfigure(0, weight=1)

        self.mainframe = ttk.Frame(self.master, padding='5 5 5 5')
        self.mainframe.grid(row=0, column=0, sticky="NEWS")
        self.mainframe.rowconfigure(0, weight=1)
        self.mainframe.columnconfigure(0, weight=1)

        self.load_file_frame = ttk.LabelFrame(self.mainframe, padding='5 5 5 5', text='Choose File')
        self.load_file_frame.grid(column=0, row=0, sticky='NEW')
        self.load_file_frame.columnconfigure(0, weight=1)
        self.load_file_frame.rowconfigure(0, weight=1)

        choose_file_label = ttk.Label(self.load_file_frame, text='Please input file you would like to analyze:')
        choose_file_label.grid(column=0, row=0, columnspan=1, sticky='NEWS')

        self.choose_file_text = ttk.Entry(self.load_file_frame, width=70, textvariable=self._filepath)
        self.choose_file_text.grid(column=0, row=1, sticky='NEWS')

        choose_file_button = ttk.Button(self.load_file_frame, text="Browse", command=self.getfilepath)
        choose_file_button.grid(row=2, column=0, sticky='W')

        load_file_button = ttk.Button(self.load_file_frame, text="Load", command=self.loadfilepath)
        load_file_button.grid(row=1, column=1, sticky='E')

        self.data_file_frame = ttk.LabelFrame(self.mainframe, padding='5 5 5 5', text='Data')
        self.data_file_frame.grid(column=0, row=1, sticky='NEW')
        self.data_file_frame.rowconfigure(1, weight=1)
        self.data_file_frame.columnconfigure(0, weight=1)

        show_plot = ttk.Button(self.data_file_frame, text='Plot Waveform', command=self.createplot)
        show_plot.grid(column=0, row=2, sticky='WN')

        show_time = ttk.Button(self.data_file_frame, text='Time', command=self.extracttime)
        show_time.grid(column=0, row=3, sticky='WN')

        show_plot2 = ttk.Button(self.data_file_frame, text='Resonant frequency ', command=self.PlotFrequency)
        show_plot2.grid(column=0, row=4, sticky='WN')

        show_plot3 = ttk.Button(self.data_file_frame, text='RT60 plot', command=self.RT60)
        show_plot3.grid(column=0, row=5, sticky='WN')

        show_plot3 = ttk.Button(self.data_file_frame, text='Frequency and Amplitude', command=self.Frequency)
        show_plot3.grid(column=0, row=6, sticky='WN')

        show_plot3 = ttk.Button(self.data_file_frame, text='RT60 Average', command=self.RT60avg)
        show_plot3.grid(column=0, row=7, sticky='WN')

        self.status_frame = ttk.Frame(self.master, relief='sunken', padding='2 2 2 2')
        self.status_frame.grid(row=1, column=0, sticky='EWS')
        self._status_msg.set('')
        status = ttk.Label(self.status_frame, textvariable=self._status_msg, anchor=W)
        status.grid(row=0, column=0, sticky='EW')

        self.highest_resonance.set('')



    def getwavdata(self, audio_file):
        wav_file = wave.open(audio_file, 'rb')
        audio_data = wav_file.readframes(wav_file.getnframes())
        return np.frombuffer(audio_data, np.int16)

    def getfilepath(self):
        self._filepath.set(tk.filedialog.askopenfilename())

    def loadfilepath(self):
        try:
            if str(self.choose_file_text.get()) != '':
                self._filepath.set(self.choose_file_text.get())
                self.sb(f'File path set as \"{self._filepath.get()}\"')
                self.converttowav(self._filepath.get())
            else:
                self.sb('File path cannot be empty.')
        except Exception as e:
            self.sb(f'The error is {e}')

    def sb(self, msg):
        self._status_msg.set(msg)

    def converttowav(self, audio_file_path):
        if not os.path.isfile(audio_file_path):
            self.sb(f"Error: File {audio_file_path} not found.")
            return

        supported_extensions = ['.mp3', '.wav', '.ogg', '.m4a']  # audio extensions
        file_extension = os.path.splitext(audio_file_path)[1].lower()

        if file_extension not in supported_extensions:
            self.sb(f"Error: Unsupported file format ({file_extension}). Supported formats are {supported_extensions}.")
            return
        else:
            try:
                audio_file = AudioSegment.from_file(
                    audio_file_path,
                    format=os.path.splitext(audio_file_path)[-1].strip('.')
                )
                wav_data = audio_file.raw_data
                self.wav_audio = AudioSegment(
                    wav_data,
                    frame_rate=audio_file.frame_rate,
                    sample_width=audio_file.sample_width,
                    channels=audio_file.channels
                )
                self.wav_audio = self.wav_audio.set_channels(1)
                self.frame_amount = len(self.wav_audio.get_array_of_samples())
                self.wav_audio.set_channels(1)
                self.raw_data = np.frombuffer(self.wav_audio.raw_data, dtype=np.int16)
                self.gethighestresonance()
            except Exception as e:
                self.sb(f"Error during conversion: {e}")

    def extracttime(self):
        if self.wav_audio is not None:
            time_duration = len(self.wav_audio) / 1000

            time_min = time_duration // 60
            time_sec = round(time_duration % 60)

            time_string = f'{time_min} minutes {time_sec} seconds'
            self.sb(f"Time is: {time_string}")
        else:
            self.sb(f'Make sure to press load')

    def createplot(self):
        if self.wav_audio is not None:
            samples = np.array(self.wav_audio.get_array_of_samples())
            # Calculate time values for x-axis
            duration_in_sec = len(self.wav_audio) / 1000
            time_values = np.linspace(0, duration_in_sec, num=len(samples))

            # Plot the waveform
            figure = Figure(figsize=(7, 4), dpi=100)
            plot = figure.add_subplot(1, 1, 1)
            plot.plot(time_values, samples)
            plot.set_xlabel("Time (seconds)")
            plot.set_ylabel("Amplitude")
            plot.set_title('Waveform of ' + self._filepath.get().split('/')[-1])

            canvas = FigureCanvasTkAgg(figure, master=self.data_file_frame)
            canvas.draw()
            canvas.get_tk_widget().grid(column=0, row=3)
        else:
            self.sb(f'Make sure to press load')

    def gethighestresonance(self):

        data = np.array(self.wav_audio.get_array_of_samples())

        frequencies, power = scipy.signal.welch(data, self.wav_audio.frame_rate, nperseg=4096)
        dominant_frequency = frequencies[np.argmax(power)]
        self.highest_resonance.set(str(dominant_frequency))


    def PlotFrequency(self):
        if self.wav_audio is not None:
            # Convert AudioSegment to NumPy array
            samples = np.array(self.wav_audio.get_array_of_samples())

            # Find the index of the maximum amplitude
            max_amplitude_index = np.argmax(np.abs(samples))

            # Calculate time corresponding to the index
            time_of_max_amplitude = max_amplitude_index / self.wav_audio.frame_rate
            self.gethighestresonance()
            self.sb(f"Highest amplitude happened at {round(time_of_max_amplitude, 2)} and the resonant frequency was {round(float(self.highest_resonance.get()), 2)}")
        else:
            self.sb(f'Make sure to press load')

    def RT60(self):
        if self.wav_audio is not None:
            sample_rate = self.wav_audio.frame_rate
            data = np.array(self.wav_audio.get_array_of_samples())
            spectrum, freqs, t, im = plt.specgram(data, Fs=sample_rate, NFFT=1024)
            plt.close()

            def find_target_frequency(freqs, target_freq):
                for x in freqs:
                    if x > target_freq:
                        break
                return x

            def frequency_check(target_freq):
                global target_frequency
                target_frequency = find_target_frequency(freqs, target_freq)
                index_of_frequency = np.where(freqs == target_frequency)[0][0]

                data_for_frequency = spectrum[index_of_frequency]

                data_in_db_fun = 10 * np.log10(data_for_frequency + 1)
                return data_in_db_fun

            def find_nearest_value(array, value):
                array = np.asarray(array)
                idx = (np.abs(array - value)).argmin()
                return array[idx]

            loop = {1: 20, 2: 1000, 3: 5000}
            self.count += 1
            self.count = self.count % 3 + 1
            data_in_db = frequency_check(loop[self.count])

            index_of_max = np.argmax(data_in_db)

            value_of_max = data_in_db[index_of_max]

            sliced_array = data_in_db[index_of_max:]
            value_of_max_less_5 = value_of_max - 5

            value_of_max_less_5 = find_nearest_value(sliced_array, value_of_max_less_5)
            index_of_max_less_5 = np.where(data_in_db == value_of_max_less_5)

            value_of_max_less_25 = value_of_max - 25
            value_of_max_less_25 = find_nearest_value(sliced_array, value_of_max_less_25)
            index_of_max_less_25 = np.where(data_in_db == value_of_max_less_25)

            rt60 = (t[index_of_max_less_5] - t[index_of_max_less_25])[0]

            seconds = np.round(abs(rt60), 3)
            self.sb(f'The RT60 reverb time at freq {int(target_frequency)} Hz is {seconds} seconds')
            self.RT60values[self.count] = seconds

            figure = Figure(figsize=(7, 4), dpi=100)
            plot = figure.add_subplot(1, 1, 1)
            plot.plot(t, data_in_db, linewidth=1, alpha=.7, color='#004bc6')
            plot.plot(t[index_of_max], data_in_db[index_of_max], 'go')
            plot.plot(t[index_of_max_less_5], data_in_db[index_of_max_less_5], 'yo')
            plot.plot(t[index_of_max_less_25], data_in_db[index_of_max_less_25], 'ro')
            plot.set_xlabel("Time (seconds)")
            plot.set_ylabel("DB")
            plot.set_title("Audio Waveform")

            canvas = FigureCanvasTkAgg(figure, master=self.data_file_frame)
            canvas.draw()
            canvas.get_tk_widget().grid(column=0, row=3)
        else:
            self.sb(f'Make sure to press load')

    def RT60avg(self):
        try:
            if self.wav_audio is not None:
                s = 0
                for x in range(3):
                    s += self.RT60values[x+1]
                avg = s / 3
                avg -= 0.5
                self.sb(f'The RT60 minus 0.5 is {round(avg, 3)}')
            else:
                self.sb(f'Make sure to press load')
        except Exception as e:
            self.sb(f"Press RT60 plot at least 3 times first")

    def Frequency(self):
        if self.wav_audio is not None:
            def calculate_frequency_spectrum(samples, frame_rate):
                # Perform Fourier Transform
                fft_result = np.fft.fft(samples)

                # Calculate the frequencies corresponding to the FFT result
                frequencies = np.fft.fftfreq(len(fft_result), 1 / frame_rate)

                # Take the absolute value to get the amplitude spectrum
                amplitudes = np.abs(fft_result)

                # Keep only the positive frequencies
                positive_frequencies = frequencies[:len(frequencies) // 2]
                positive_amplitudes = amplitudes[:len(amplitudes) // 2]

                return positive_frequencies, positive_amplitudes

            samples = np.array(self.wav_audio.get_array_of_samples())
            frequencies, amplitudes = calculate_frequency_spectrum(samples, self.wav_audio.frame_rate)

            # Filter frequencies within the specified range
            mask = np.logical_and(frequencies >= 20, frequencies <= 20000)
            filtered_frequencies = frequencies[mask]
            filtered_amplitudes = amplitudes[mask]

            # Plot the specific frequencies
            figure = Figure(figsize=(7, 4), dpi=100)
            plot = figure.add_subplot(1, 1, 1)
            plot.plot(filtered_frequencies, filtered_amplitudes)

            plot.set_xlabel("Frequency (Hz)")
            plot.set_ylabel("Amplitude")
            plot.set_title(f"Frequency Spectrum ({20} Hz to {20000} Hz)")

            canvas = FigureCanvasTkAgg(figure, master=self.data_file_frame)
            canvas.draw()
            canvas.get_tk_widget().grid(column=0, row=3)
        else:
            self.sb(f'Make sure to press load')

if __name__ == "__main__":
    root = Tk()
    app = AudioAnalyzerApp(root)
    root.mainloop()
